Adds ApiStop and StopMapper in response to #2968

bardrotzer opened the pull request:
In response to bug #2968 

I've tried to implement the recommendations from @t2gran with a small exception. The recommendation was to add the `StopMapper` to the `GraphIndex`, i guess this was a typo, as the `GraphIndex` class exposes the models that are being consumed by the main application. I have added  `StopMapper`, `ApiStops` and `ApiStopShort` to the `IndexAPI` to prevent the `Stop` model  from being exposed to the Api directly. 

There are issues with the API that are not addressed here, and there are instances of the `Stop` Model that are still in use as they are consumed by other classes (`Route`, `TripPattern`, `StopTimesInPattern`). Let me know if we should refactor and update the model being consumed by these.

Not being a Java developer by day, I'm hesitant to make any architectural decisions, but happy to help wherever refactoring is required. Could it be an idea to split the requests in `IndexApi` into separate Resource classes (as is done with the `PlannerResource`)?


- [x] **issue**: Link to or create an [issue](https://github.com/opentripplanner/OpenTripPlanner/issues) that describes the relevant feature or bug. Add [GitHub keywords](https://help.github.com/articles/closing-issues-using-keywords/) to this PR's description (e.g., `closes #45`).
- [x] **roadmap**: Check the [roadmap](https://github.com/orgs/opentripplanner/projects/1) for this feature or bug. If it is not already on the roadmap, PLC will discuss as part of the review process.
- [ ] **tests**: Have you added relevant test coverage? Are all the tests passing on [the continuous integration service (Travis CI)](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#continuous-integration)?
- [x] **formatting**: Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#code-style)? 
- [ ] **documentation**: If you are adding a new configuration option, have you added an explanation to the [configuration documentation](docs/Configuration.md) tables and sections?
- [ ] **changelog**: add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue

To be completed by @opentripplanner/plc:

- [ ] reviews and approvals by 2 members, ideally from different organizations
- [ ] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

t2gran commented on src/main/java/org/opentripplanner/index/IndexAPI.java:
```diff
@@ -153,14 +156,14 @@ public Response getAgencyRoutes (@PathParam("feedId") String feedId, @PathParam(
    public Response getStop (@PathParam("stopId") String stopIdString) {
        FeedScopedId stopId = GtfsLibrary.convertIdFromString(stopIdString);
-       Stop stop = index.stopForId.get(stopId);
+       ApiStop stop = StopMapper.mapStop(index.stopForId.get(stopId));
```
This was returning a "domain" Stop before, I cant see that the StopAdapter comes into play here. Or, did I miss something?

It seems like there is a lot of types in play for just stop on this API, but if the "domain" stop is retuned then we need to make an API type for the old `dev-1.x` Stop, too. I can discuss this @abyrd and possible the community - maybe it is better to clean it up.

t2gran replied:
@abyrd What do you think about this. I think this is a mistake, the old `StopType` (now `ApiStop`) have the same field as the domain `Stop`, but with slightly different names. I think we should just return the `ApiStop` here, and not create another `ApiStop2` to be backward compatible.

t2gran replied:
I apologize for this one, I managed to confuse my self - with all the different Stop objects in. use. I think what you did here is right.

bardrotzer commented:
@t2gran I'm sorry I've been a bit absent from the project lately, but picking it up and looking at it now. Your comments concering the *domain* and *api* Stops are correct, I refactored to keep it cleaner.

As I moved on, it seems to me that I might have misunderstood the refactoring request, but as I understood it, you wanted to ensure that the `Stop` model was not exposed on the `IndexApi`. 
> the Stop class should not be exposed on the REST API 

and that we should replace `Stop` with the  `ApiStop` in the `GraphIndex` mappings

> call the mapper from GraphIndex

I did a test where this was implemented, but the side effects are quite big and requires more refactoring than I am comfortable with. I can do the coding, but replacing the `Stop`model with a mapped `ApiStop` will affect `TripPattern`which will affect `StopPattern` and so on.

As of now, I do not feel comfortable making these changes. Do feel free to discuss this. I'm more than happy to help if anything is needed.

t2gran commented:
Thank you for the work so fare. This is indeed confusing, having 2 kind of Stop types retuned on the Index API -  it might even be a mistake. I was not aware that the Index API was returning the domain Stop before reading the review. I am aware that the domain model do leek out in various other places and the goal is to fix it. 

It is better to keep PRs relatively small with a lot of parallell work going on, so I think we can leave the domain `Stop` and `TripPattern` for now.

t2gran commented:
This PR is no longer relevant, the PR #3031 refactor the entire API classes and clean up the REST API. I am sorry for the wasted efford. I did look at incorporating this into the #3031, but it turned out to be easier to just do it. I will close this now.
